File: 2022-2023/sourcecode/lab/lab13/OfflineCachingBasicsJava/app/src/main/java/pl/udu/uwr/pum/offlinecachingbasicsjava/util/NetworkBoundResource.java
package pl.udu.uwr.pum.offlinecachingbasicsjava.util;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import pl.udu.uwr.pum.offlinecachingbasicsjava.data.db.UserDatabase;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public abstract class NetworkBoundResource<ResultType, RequestType> {
    private final MediatorLiveData<Resource<ResultType>> result = new MediatorLiveData<>();

    protected NetworkBoundResource() {
        result.setValue(Resource.loading(null));
        LiveData<ResultType> dbSource = loadFromDb();
        result.addSource(dbSource, data -> {
            result.removeSource(dbSource);
            if (shouldFetch(data)) {
                fetchFromNetwork(dbSource);
            } else {
                result.addSource(dbSource, newData -> result.setValue(Resource.success(newData)));
            }
        });
    }

    private void fetchFromNetwork(final LiveData<ResultType> dbSource) {
        result.addSource(dbSource, newData -> result.setValue(Resource.loading(newData)));
        createCall().enqueue(new Callback<RequestType>() {
            @Override
            public void onResponse(@NonNull Call<RequestType> call, @NonNull Response<RequestType> response) {
                result.removeSource(dbSource);
                saveResultAndReInit(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<RequestType> call, @NonNull Throwable t) {
                onFetchFailed();
                result.removeSource(dbSource);
                result.addSource(dbSource, newData -> result.setValue(Resource.error(t.getMessage(), newData)));
            }
        });
    }

    private void saveResultAndReInit(RequestType response) {
        UserDatabase.databaseWriteExecutor.execute(() -> saveCallResult(response));
        result.addSource(loadFromDb(), newData -> result.setValue(Resource.success(newData)));

    }

    protected abstract void saveCallResult(@NonNull RequestType item);

    protected boolean shouldFetch(@Nullable ResultType data) {
        return true;
    }

    @NonNull
    protected abstract LiveData<ResultType> loadFromDb();

    @NonNull
    protected abstract Call<RequestType> createCall();

    protected void onFetchFailed() {
    }

    public final LiveData<Resource<ResultType>> getAsLiveData() {
        return result;
    }
}

File: 2022-2023/sourcecode/lab/lab13/PolishNewsApp/app/src/main/java/pl/udu/uwr/pum/polishnewsapp/util/NetworkBoundResource.kt
package pl.udu.uwr.pum.polishnewsapp.util

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

inline fun <ResultType, RequestType> networkBoundResource(
    crossinline query: () -> Flow<ResultType>,
    crossinline fetch: suspend () -> RequestType,
    crossinline saveFetchResult: suspend (RequestType) -> Unit,
    crossinline shouldFetch: (ResultType) -> Boolean = {true},
    crossinline fetchFailed: (Throwable) -> Unit = {},
    crossinline fetchSuccess: () -> Unit = {}
) = channelFlow {
    val data = query().first()
    if (shouldFetch(data)){
        val loading = launch { query().collect{send(Resource.Loading(it))} }
        try {
            saveFetchResult(fetch())
            fetchSuccess()
            loading.cancel()
            query().collect { send(Resource.Success(it)) }
        } catch (t: Throwable) {
            fetchFailed(t)
            loading.cancel()
            query().collect{send(Resource.Error(t, it))}
        }
    } else query().collect{send(Resource.Success(it))}
}